### DISCUSSION TITLE
Welcome to Shapiro Discussions!

### DISCUSSION THREAD

**Discussion by mathiasrichter** [mathiasrichter/shapiro#15] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by brownjasonj** [comment]
I've been looking at the ld+json response for a simple example (below).  I have defined a property and shape for a string property with a minimum length 1 and maximum length of 35.  I want to reuse this property/shape for other properties, so i use subproperty.  You see that two properties are defined in this way as part of the Class Item.

When I get ld+json of the Item class there are no type definitions which I would have expected, however perhaps i'm not using the property/propertyshape and subproperty correctly.  Would be great to understand this better and get support for this in Shapiro.

@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .

@prefix example: <http://127.0.0.1:8000/simple/> .


example:minmaxTextProperty a rdfs:Property, sh:PropertyShape;
   sh:datatype xsd:string;
   sh:minLength 1;
   sh:maxLength 35.

example:ItemFormalName a rdf:Property, sh:PropertyShape;
   rdfs:label "ItemFormalName";
   rdfs:description "The name of an item";
   sh:subPropertyOf example:minmaxTextProperty.

example:ItemNickName a rdf:Property, sh:PropertyShape;
   rdfs:label "ItemNickName";
   rdfs:description "The nick name of an item";
   sh:subPropertyOf example:minmaxTextProperty.

example:Item a rdfs:Class, sh:NodeShape;
   rdfs:label "Item";
   rdfs:description "An Item has two properties ";
   sh:property [
      sh:path example:ItemFormalName;
      sh:maxCount 1;
   ];
   sh:property [
      sh:path example:ItemNickName;
      sh:maxCount 1;
   ].

**Comment by mathiasrichter** [comment]
Thanks for this use case - this was interesting to research and I've learned the following about SHACL: SHACL by default does not infer anything from declarations of rdfs:subPropertyOf (you're using sh:subPropertyShapeOf in your example, which I don't think exists in the SHACL spec, at least I did not find it).

Shapiro needs to explicitly traverse any rdfs:subPropertyOf declarations and include the constraints inherited from "super-properties". 

I have opened issue #27 and will aim to implement as soon as possible.

**Reply by brownjasonj** [reply]
Thanks for looking into this.  i've fixed my example to use subPropertyOf.  You say that SHACL doesn't infer anything via rdfs:subPropertyOf, which makes sense since it is Shacl.  So, my the question would be does SHACL have any inheritance capabilities so that you can create hierarchies of constraints.....as the example has?

**Comment by mathiasrichter** [comment]
Shacl infers the use of rdfs:subClassOf during validation, so it will ensure that validation for a NodeShape with sh:targetClass x will also validate all NodeShapes of all superclasses of x. There is an inheritance example in the codebase here: https://github.com/mathiasrichter/shapiro/blob/main/test/ontologies/com/example/org/nodeshapes_inheritance.ttl

**Comment by brownjasonj** [comment]
Mat, i found another possible way to encode the above example, see below.  I am defining a new datatype as a subclass of xsd:string and putting a shacl nodeshape constraint on that datatype.

It seems more logical as an approach, but would be interested in your thoughts.  Incidently requesting the ld+json for the class sample2/Item returns a json schema without types for the two properties so Shapiro doesn't traverse the subclass definitions.

@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .

@prefix sample2: <http://127.0.0.1:8000/simple2/> .


sample2:minmaxTextType a rdfs:DataType, sh:PropertyShape;
   rdfs:subClassOf xsd:string;
   sh:minLength 1;
   sh:maxLength 35.

sample2:ItemFormalName a rdf:Property, sh:PropertyShape;
   rdfs:label "ItemFormalName";
   rdfs:description "The name of an item";
   sh:datatype sample2:minmaxTextType.

sample2:ItemNickName a rdf:Property, sh:PropertyShape;
   rdfs:label "ItemNickName";
   rdfs:description "The nick name of an item";
   sh:datatype sample2:minmaxTextType.

sample2:Item a rdfs:Class, sh:NodeShape;
   rdfs:label "Item";
   rdfs:description "An Item has two properties ";
   sh:property [
      sh:path sample2:ItemFormalName;
      sh:maxCount 1;
   ];
   sh:property [
      sh:path sample2:ItemNickName;
      sh:maxCount 1;
   ].
